Road to Tor v3 support (BIP155)

furszy opened the pull request:
Conjunction of a large number of back ports, updates and refactorings that made with the final goal of implementing v3 Onion addresses support (BIP155 https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki) before the tor v2 addresses EOL, scheduled, by the Tor project, for (1) July 15th: v2 addr support removal from the code base, and (2) October 15th: v2 addr network disable, where **every peer in our network running under Tor will loose the connection and drop the network**.

As BIP155 describes, this is introducing a new P2P message to gossip longer node addresses over the P2P network. This is required to support new-generation Onion addresses, I2P, and potentially other networks that have longer endpoint addresses than fit in the 128 bits of the current addr message.

In order to achieve the end goal, had to:
1.  Create Span class and push it up to latest Bitcoin implementation. 
2.  Update the whole serialization framework and every object using it up to latest Bitcoin implementation (3-4 years ahead of what we currently are in master).
3.  Update the address manager implementing ASN-based bucketing of the network nodes.
4.  Update and refactor the netAddress and address manager tests to latest Bitcoin implementation (4 years ahead of what we currently are in master).
5.  Several util string, vector, encodings, parsing, hashing backports and more..

Important note: 
This PR it is not meant to be merged as a standalone PR, will decouple smaller ones moving on. Adding on each sub-PR its own description isolated from this big monster.

Second note:
This is still a **work-in-progress**, not ready for testing yet. I'm probably missing to mention few PRs that have already adapted to our sources. Just making it public so can decouple the changes, we can start merging them and i can continue working a bit more confortable (rebase a +170 commits separate branch is not fun..).

### List of back ported and adapted PRs:

Span and Serialization:
----------------
*  #12886.
*  #12916.
*  #13558.
*  #13697. (Only Span's commit 29943a90)
*  #17850.
*  #17896.
*  #12752.
*  #16577.
*  #16670. (without faebf627)
*  #17957.
*  #18021.
*  #18087.
*  #18112 (only from 353f3762 that we don't support).
*  #18167.
*  #18317.
*  #18591 (only Span's commit 0fbde488)
*  #18468.
*  #19020.
*  #19032.
*  #19367.
*  #19387.

Net, NetAddress and AddrMan:
----------------

*  #7932.
*  #10756.
*  #10765.
*  #12218.
*  #12855.
*  #13532.
*  #13575.
*  #13815.
*  #14532.
*  #15051.
*  #15138.
*  #15689.
*  #16702.
*  #17243.
*  #17345.
*  #17754.
*  #17758.
*  #17812.
*  #18023.
*  #18454.
*  #18512.
*  #19314.
*  #19687

Keys and Addresses encoding:
----------------
* #11372.
* #17511.
* #17721.

Util:
----------------
* #9140.
* #16577.
* #16889.
* #19593.

Bench:
----------------
* #16299.

BIP155:
----------------
*  #19351.
*  #19360.
*  #19534.
*  #19628.
*  #19841.
*  #19845.
*  #19954.
*  #19991 (pending).
*  #19845.
*  #20000 (pending).
*  #20120.
*  #20284.
*  #20564.
*  #21157 (pending).
*  #21564 (pending).
*  Fully removed v2 onion addr support.
*  Add hardcoded seeds.
*  Add release-notes, changes to files.md and every needed documentation.

I'm currently working on the PRs marked as "pending", this isn't over, but I'm pretty pretty close :). What a long road..

eval9 commented:
will this solve https://github.com/PIVX-Project/PIVX/issues/2454 ?  (currently unable to use tor/onion addresses with master nodes (tor ver 2.0)

furszy commented:
Not related, this is for Tor v3 onion addresses support. Just answered you there.

furszy commented:
Rebased on master after decoupled PRs merge + squashed some commits. From 179 commits to 132 now :). Next in the line is #2412 🤘.

furszy commented:
rebased after #2422 merge.
 Next one in line is #2470, plus will decouple all the ASN-based IP bucketing work into another PR, we can go ahead with it now 🤘.

furszy commented:
Rebased, conflicts and lint-python warnings fixed.

The remaining open PRs decoupled from this work are:
* [ ] #2470.
* [ ] #2479.
* [ ] #2480.
* [ ] #2464.

Once they are all merged, this PR will only contain the Tor v3 onion addrs node & network support.

furszy commented:
rebased on master, 68 commits left. Next one in the line is #2480.

furszy commented:
Had to add four others back ports to solve the linux-only `rpc_bind` test issue:
* #12482.
* #12510.
* #14790.
* #14861.

furszy commented:
rebased, next one in the line is #2480.

furszy commented:
Added upstream's pending PRs 19961, 20120, 20284, 21753 and 22172 for this work.

furszy commented:
Rebased, GA passing nicely, ready for review.

furszy commented:
rebased on master so it gets the `assert_start_raises_init_error` code move.

random-zebra commented on test/functional/test_framework/util.py:
```diff
@@ -322,6 +322,12 @@ def initialize_datadir(dirname, n):
 
+def append_config(dirname, n, options):
+    datadir = get_datadir_path(dirname, n)
+    with open(os.path.join(datadir, "bitcoin.conf"), 'a', encoding='utf8') as f:
```
(32cc3fa349b56affb64b4a96c798228cdd9de083)

```suggestion
    with open(os.path.join(datadir, "pivx.conf"), 'a', encoding='utf8') as f:
```

random-zebra reviewed (commented):
What a long road... :coffee: Code review ACK. Will do some more testing.

furszy commented:
updated per feedback

Fuzzbawls commented on src/protocol.cpp:
```diff
@@ -19,6 +19,8 @@ namespace NetMsgType
 const char* VERACK = "verack";
 const char* ADDR = "addr";
+const char *ADDRV2="addrv2";
+const char *SENDADDRV2="sendaddrv2";
```
nit: pointer/reference on the left side as per clang-format rules:
```cpp
const char* ADDRV2="addrv2";
const char* SENDADDRV2="sendaddrv2";
```

furszy replied:
changed

Fuzzbawls commented on src/qt/pivx/topbar.cpp:
```diff
@@ -608,8 +608,7 @@ void TopBar::updateTorIcon()
         }
-        QString ip_port_q = QString::fromStdString(ip_port);
-        ui->pushButtonTor->setButtonText(tr("Tor Active: %1").arg(ip_port_q));
+        ui->pushButtonTor->setButtonText(tr("Tor Active"));
```
could we maybe use `setToolTip()` to display the onion address? also, maybe a future PR could make this button an action button that allows for the onion address to be copied to the clipboard?

furszy replied:
sure, will add the tooltip.

I left a comment in the commit: ideally, a future good work would be to create a network-only information dialog, grouping and presenting the network/proxy related information properly (networks, addresses, reachability, if it's a limited net, etc). So we keep the topbar menu buttons as essentially actionables for sub-menus and/or other information screens/dialogs.

Still, atm, we know that no regular user will really use it. Something for the future when there is some free time to spend on extra things.

Fuzzbawls commented on test/functional/p2p_addr_relay.py:
```diff
@@ -0,0 +1,70 @@
+from test_framework.util import (
+    assert_equal,
+)
+import time
```
nit: `import time` should be above any local imports and separated with an empty line

furszy replied:
added, will be good to add the rule to the python-lint script.

Fuzzbawls commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -3,13 +3,18 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test node responses to invalid network messages."""
+import struct
```
nit: empty line between the docstring and the first import

furszy replied:
will be good to add the rule to the python-lint script.

Fuzzbawls reviewed (commented):
Initial review and light testing: left some nit/styling comments that are non-functional-changing. Have so far tested that the client receives a proper v3 onion address (when the tor router is properly configured) via the control socket. next step is to test node interoperability using these v3 addresses.

furszy commented on doc/release-notes.md:
```diff
@@ -277,6 +284,19 @@ It enforces the same rules as the legacy cold-staking opcode, but without allowi
+  `-listenonion` configuration parameter will now be created as a Tor v3 service
+  instead of Tor v2. The private key that was used for Tor v2 (if any) will be
+  left untouched in the `onion_private_key` file in the data directory (see
+  `-datadir`) and can be removed if not needed. Bitcoin Core will no longer
```
changed

furszy commented:
update per feedback. Squashes the nits on their respective commits. Easier way of checking it is with a `git range-diff`

random-zebra reviewed (dismissed):
ACK 94f82ec7561c593c8f4b83dafd8af62cbf7b0bcd


> Easiest way of checking it is with a `git range-diff`

Or with a simple two-dots diff :) https://github.com/PIVX-Project/PIVX/compare/2aa1ccc414046685e3f897d5de4ee3b6cc65886a..94f82ec7561c593c8f4b83dafd8af62cbf7b0bcd :exploding_head:

furszy commented:
yep cool, that is even simpler.

Once this one gets merged, will rapidly add the v5.3 enforcement height for testnet so we can bump it and start the release process with what we have in master + 2492 (the rest can be added in a subsequent point release).
Time is moving fast and we are getting pretty close to the release window.

Fuzzbawls commented:
Need to add https://github.com/bitcoin/bitcoin/pull/20129 to this, otherwise gitian builds will fail the symbol check.

Also having an issue with starting a MN that uses a TORv3 address, the MN daemon crashes shortly after receiving the MN broadcast from the controller (both using TORv3 connectivity):
```bash
pivxd: netaddress.cpp:1055: bool CSubNet::Match(const CNetAddr&) const: Assertion `network.m_addr.size() == addr.m_addr.size()' failed.
```

furszy commented:
added btc@20129 👍 .

About the assertion initializing a MN, strange. i just started it via the startup args and using the RPC `initmasternode` command and both actions worked properly (both running 70923 locally).

Side thing, unrelated to this assertion, i'm adding a guard to the mnb message creation & processing for the new addr v2 serialization. So upgraded peers don't create nor broadcast an mnb containing a new addr v2 to non-upgraded peers before the new NU enforcement (introduced in #2492). Will push it soon.

furszy commented:
Have made few changes and tested this last version the whole day and all has been running properly. Locally in OSX (regtest and testnet) and even have set up a new MN on linux as well from scratch (ubuntu 18.04.1 LTS, running an old tor version 0.3.2.10 that is the latest provided by apt-get).

Have left my VPS running a testnet MN on `x4ys3tklrydij35o7m7x4f2baoolmew3q45r7dvlpdkzuoogfb3lbfqd.onion`. Feel free to connect and play with it if you want (you will need to drop c0c0d5d and add the peer connection manually).
Remember that non-upgraded peers are not able to decode the mnb with the new addr serialization.

Side from all of that, the best way of testing this further is with a bumped network and #2492. Non-upgraded peers will not receive and there by will not propagate v2 addresses across the network.

Let me know how it goes there.

furszy commented:
Rebased after 2492 merge, added enforcement height for testnet (for block 332,300) and zebra's be46655 for the spork dance.

As we are running out of time (only 7 days left until the next superblock), would be better to leave every other PR tagged for v5.3 that does not contain a consensus change or a bugfix for a subsequent v5.4 (that we can release a week or two after v5.3 without any rush. Mostly for #2280 and #2406 that will need some more work and further proper testing) and only be focused on testing what is here and in master currently (more than +150 PRs..) so we can pack it on time.

random-zebra commented on src/masternode.cpp:
```diff
@@ -293,6 +294,13 @@ bool CMasternodeBroadcast::Create(const std::string& strService,
     if (!CheckDefaultPort(_service, strErrorRet, "CMasternodeBroadcast::Create"))
         return false;
 
+    // Check if the MN has a ADDRv2 and reject it if the new NU wasn't enforced.
+    if (!_service.IsAddrV1Compatible() &&
+        !Params().GetConsensus().NetworkUpgradeActive(chainHeight, Consensus::UPGRADE_V5_3)) {
+        strErrorRet = "Cannot start MN with a v2 address before the v5.3 enforcement\n. Let the node fully sync before start the masternode";
+        return false;
+    }
```
Nit: The second line (`Let the node fully sync before start the masternode`) results cropped and hard to read in the GUI `SnackBar`. 
It is also redundant, as the block-sync is already checked for at line 266, so the node is synced here. So I think we can just drop it (besides, it should be "before start**ing**" imo).

furszy replied:
sure, pushed.

random-zebra reviewed (approved):
tested running v3 onion masternode on testnet (removing 2feabac) all good.
tested gitian build (merge be46655fcd70b5e458433826d40147631ec21501 on acdbcf8420ebf9c6fa972cc1c3b3d51b134b2c9a)
```
d5a129376b55a3f2b5ee68f60a577b828eab605dfef7d84d3716a58b0f884981  pivx-5.2.99-aarch64-linux-gnu-debug.tar.gz
b752241e653e3d7075df326e5557efef82490962f1b5e69ddf03b96b0b2eb4c8  pivx-5.2.99-aarch64-linux-gnu.tar.gz
b2d0c6130c8feb699d1459aebb7f24587ced5fc24dfe3730ca7667d5a88b426b  pivx-5.2.99-arm-linux-gnueabihf-debug.tar.gz
90ff8abd1ac1d49736194071582bcb3bbf39bf6bb9fddec84289da8a19dc7459  pivx-5.2.99-arm-linux-gnueabihf.tar.gz
2246d21135ace3b8a61562e05250fdad676766f466f79e0af9561dda437929cb  pivx-5.2.99-i686-pc-linux-gnu-debug.tar.gz
f59dd658993ed83fef81b80fe22e609994103726ae6bf761a6f16f10e9c9ea0a  pivx-5.2.99-i686-pc-linux-gnu.tar.gz
6b7de40df7364f4a6ca4a9850c78bc91022e88337deee9496b10a6e93a8b65ab  pivx-5.2.99-x86_64-linux-gnu-debug.tar.gz
0543f47b5ca9d27bbf88c9bfd0331d974e9b3d060d254fcf4e7b254c6ac778fa  pivx-5.2.99-x86_64-linux-gnu.tar.gz
41932488bb747edbb89bbe0f938f6458ae9b3e2105270184c6c6a3e5624fa774  src/pivx-5.2.99.tar.gz
323e92f7e1cd81c04ced477dda950ecd5ccaa543bbb54d1ae834c1d0a6874028  pivx-linux-5.2-res.yml

1f08c53ca11292ebb2ed0d70bbff6f45ee0bf5462da05752fb7d42c49a81dc71  pivx-5.2.99-win-unsigned.tar.gz
32c1bb799871b3582a8a7874752b61b234ca7b074911ff18fe0dfbc070db224c  pivx-5.2.99-win64-debug.zip
909c7f9311ef830ba362ff25daa64051b860711bc9f22822a9a93e7e9f3b36e1  pivx-5.2.99-win64-setup-unsigned.exe
902fe60b9b20c4f022aa325a3771e8227aec15366ac1fccade557a79c877a4b2  pivx-5.2.99-win64.zip
582312239f815b41381283c1c6fee1c32e4312385548d2e8c9f9a159d69ba696  src/pivx-5.2.99.tar.gz
1afadc029ab2c9c9087e827ebfbea394824bd2a50c039419a51987e9708c44ed  pivx-win-5.2-res.yml

d1606c66356bee93e8180331a18574b3b0ea42eb5cfc23373b16ae68c8ff91ba  pivx-5.2.99-osx-unsigned.dmg
140e28f5a31eb8c386706c76ad485a5d7db9aa437434e8f85b857b51d97a4cad  pivx-5.2.99-osx-unsigned.tar.gz
fa64a0064d47193c6d162f49b7c3a946f23cc84eae4986356672e10d1f800cbc  pivx-5.2.99-osx64.tar.gz
c204f92e39afd8bd357f698199cfc332823fdbfc99402e7e862524cc8d650b48  src/pivx-5.2.99.tar.gz
a132f9d4a4fb891f1b434c346fc5aba9786f2a678c19f0e7c9af41dcbb2150b0  pivx-osx-6.0-res.yml
```

ACK be46655fcd70b5e458433826d40147631ec21501

random-zebra reviewed (approved):
utACK ecde04af3455e1b803ae93bea059f72915a1765d

NoobieDev12 commented:
Post-merge comment, just confirming that I was able to run it yesterday on testnet as well, it works perfectly. I only had the permission issues which were solved with the @furszy's guidance, but these issues are not related to this PR. Nice work!
